### PR TITLE
DAOS-16211 vos: Ensure we delete exact entry

### PR DESCRIPTION
**Pull request by jolivier23**
When the change was made to allow partial overwrite for rebuild, it broke delete such that delete would remove the newest extent rather than the exact one we requested.

Also, don't ignore errors when processing removals

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/vos/evtree.c [hunk 1]
```diff
@@ -170,7 +170,7 @@ time_cmp(uint64_t t1, uint64_t t2, int *out)
 		*out = RT_OVERLAP_OVER;
 	} else {
 		*out = RT_OVERLAP_UNDER;
-		if (t2 == EVT_REBUILD_MINOR_MIN) {
+		if (!exact && t2 == EVT_REBUILD_MINOR_MIN) {
```
Comment (NiuYawei) [inline src/vos/evtree.c:173]: Though the epoch should be much larger than EVT_REBUILD_MINOR_MIN (when there is no unexpected bug), I think this check should be applied to minor epoch only.

File: src/vos/evtree.c [hunk 2]
```diff
@@ -2639,8 +2639,8 @@ evt_ent_array_fill(struct evt_context *tcx, enum evt_find_opc find_opc,
 			if (find_opc == EVT_FIND_OVERWRITE)
 				has_agg = true;
 
-			evt_rect_overlap(&rtmp, rect, &range_overlap,
-					 &time_overlap);
+			evt_rect_overlap(&rtmp, rect, &range_overlap, &time_overlap,
+					 find_opc == EVT_FIND_SAME);
```
Comment (NiuYawei) [inline src/vos/evtree.c:2643]: I think the rebuild minor epoch check should only be applied when find_opc == EVT_FIND_OVERWRITE? Why do we need to apply it to EVT_FIND_ALL & EVT_FIND_FIRST?

File: src/vos/vos_aggregate.c [hunk 3]
```diff
@@ -1524,6 +1524,10 @@ insert_segments(daos_handle_t ih, struct agg_merge_window *mw, bool last, unsign
 
 	/** Remove processed removal records */
 	rc = process_removals(mw, oiter, &mw->mw_rmv_ents, last, true);
+	if (rc != 0) {
+		DL_ERROR(rc, "Unable to process extent removals");
+		goto abort;
+	}
```
Comment (NiuYawei) [inline src/vos/vos_aggregate.c:1530]: Could you explain a bit why it this rebuild overwrite change could affect removal records?

Comment (jolivier23) [reply src/vos/vos_aggregate.c:1530]: It can't but this change can avoid an assertion failure.

Comment (NiuYawei) [reply src/vos/vos_aggregate.c:1530]: Should we add a new assert in process_removals() (like asserting evt_delete() won't return -DER_ENOENT) or somewhere else?

**Review by NiuYawei — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Comment by shimizukko** [comment]
@jolivier23 Could you provide the patch on release/2.6?

**Comment by jolivier23** [comment]
> @jolivier23 Could you provide the patch on release/2.6?

https://github.com/daos-stack/daos/pull/14815

**Pull request merged**